File: packages/sdk-node/bench/throughput.mjs
// Real throughput benchmark for RateLimiter.increment(), mirroring the two
// scenarios in sdk-go/sharded_limiter_test.go (hot key, many keys). Node has
// no real thread parallelism for this in-process operation — the honest
// number here is single-threaded ops/sec, the actual ceiling a Node process
// hits under load.
//
// Run: node bench/throughput.mjs

import { RateLimiter } from '../dist/esm/core/rate-limiter.js';

function bench(name, keys, iterations) {
  const limiter = new RateLimiter({ clock: { now: () => Date.now() }, capacity: 50_000 });
  const options = {
    requestsPerSecond: 1_000_000,
    burst: 1_000_000,
    windowMs: 60_000,
    remoteRateLimitEndpoint: '',
    apiKey: undefined,
  };
  const keySet = Array.from({ length: keys }, (_, i) => `tenant-${i}`);

  const start = process.hrtime.bigint();
  for (let i = 0; i < iterations; i++) {
    limiter.increment(keySet[i % keys], options, 1);
  }
  const end = process.hrtime.bigint();

  const totalNs = Number(end - start);
  const nsPerOp = totalNs / iterations;
  const opsPerSec = 1e9 / nsPerOp;
  console.log(`${name.padEnd(24)} ${nsPerOp.toFixed(1).padStart(10)} ns/op   ${Math.round(opsPerSec).toLocaleString('en-US').padStart(14)} ops/sec`);
}

console.log(`node ${process.version}`);
bench('HotKey', 1, 2_000_000);
bench('ManyKeys (1024)', 1024, 2_000_000);

File: packages/sdk-python/bench/throughput.py
"""Real throughput benchmark for RateLimiter.increment(), mirroring the two
scenarios in sdk-go/sharded_limiter_test.go (hot key, many keys). CPython's
GIL means concurrent threads don't add real parallelism for this operation —
the honest number here is single-threaded ops/sec, the actual ceiling one
Python process hits under load.

Run: python3 bench/throughput.py
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from rateguard import RateLimiter
from rateguard.types import RateLimitOptions


class SystemClock:
    def now(self) -> float:
        return time.time() * 1000


def bench(name: str, keys: int, iterations: int) -> None:
    limiter = RateLimiter(SystemClock(), capacity=50_000)
    options = RateLimitOptions(requests_per_second=1_000_000, burst=1_000_000)
    key_set = [f"tenant-{i}" for i in range(keys)]

    start = time.perf_counter_ns()
    for i in range(iterations):
        limiter.increment(key_set[i % keys], options, 1.0)
    end = time.perf_counter_ns()

    total_ns = end - start
    ns_per_op = total_ns / iterations
    ops_per_sec = 1e9 / ns_per_op
    print(f"{name:<24} {ns_per_op:>10.1f} ns/op   {ops_per_sec:>14,.0f} ops/sec")


if __name__ == "__main__":
    print(f"python {sys.version.split()[0]}")
    bench("HotKey", 1, 500_000)
    bench("ManyKeys (1024)", 1024, 500_000)
